src/server.rs

```diff
@@ -532,15 +530,114 @@ impl WalletService for Arc<crate::wallet::Wallet> {
     async fn create_bmm_critical_data_transaction(
         &self,
         request: tonic::Request<CreateBmmCriticalDataTransactionRequest>,
     ) -> std::result::Result<tonic::Response<CreateBmmCriticalDataTransactionResponse>, tonic::Status>
     {
-        Err(tonic::Status::new(
-            tonic::Code::Unimplemented,
-            "not implemented",
-        ))
+        let CreateBmmCriticalDataTransactionRequest {
+            sidechain_id,
+            value_sats,
+            height,
+            critical_hash,
+            prev_bytes,
+        } = request.into_inner();
+
+        let amount = value_sats
+            .ok_or_else(|| missing_field::<CreateBmmCriticalDataTransactionRequest>("value_sats"))
+            .map(bdk::bitcoin::Amount::from_sat)
+            .map_err(|_| {
+                invalid_field_value::<CreateBmmCriticalDataTransactionRequest>(
+                    "value_sats",
+                    &value_sats.unwrap_or_default().to_string(),
+                )
+            })?;
+
+        let locktime = height
+            .ok_or_else(|| missing_field::<CreateBmmCriticalDataTransactionRequest>("height"))
+            .map(bdk::bitcoin::absolute::LockTime::from_height)?
+            .map_err(|_| {
+                invalid_field_value::<CreateBmmCriticalDataTransactionRequest>(
+                    "height",
+                    &height.unwrap_or_default().to_string(),
+                )
+            })?;
+
+        let sidechain_number = sidechain_id
+            .ok_or_else(|| missing_field::<CreateBmmCriticalDataTransactionRequest>("sidechain_id"))
+            .map(SidechainNumber::try_from)?
+            .map_err(|_| {
+                invalid_field_value::<CreateBmmCriticalDataTransactionRequest>(
+                    "sidechain_id",
+                    &sidechain_id.unwrap_or_default().to_string(),
+                )
+            })?;
+
+        if !self
+            .is_sidechain_active(sidechain_number)
+            .await
+            .unwrap_or_default()
+        {
+            return Err(tonic::Status::failed_precondition(
+                "sidechain is not active",
+            ));
+        }
+
+        // This is also called H*
+        let critical_hash = critical_hash
+            .ok_or_else(|| {
+                missing_field::<CreateBmmCriticalDataTransactionRequest>("critical_hash")
+            })?
+            .decode_tonic::<CreateBmmCriticalDataTransactionRequest, _>("critical_hash")
+            .map(bdk::bitcoin::BlockHash::from_byte_array)?;
+
+        let prev_bytes = prev_bytes
+            .ok_or_else(|| missing_field::<CreateBmmCriticalDataTransactionRequest>("prev_bytes"))?
+            .decode_tonic::<CreateBmmCriticalDataTransactionRequest, _>("prev_bytes")
+            .map(bdk::bitcoin::BlockHash::from_byte_array)?;
+
+        let mainchain_tip = self
+            .get_mainchain_tip()
+            .await
+            .map_err(|err| tonic::Status::internal(err.to_string()))?;
```

Comment:
You should probably log an error here, if you are not including an error trace in the status message